File: apps/mobile/ios/Sources/FlagshipUI/Shell/BiometricLockScreen.swift
import SwiftUI
import Flagship
import FlagshipCore

/// B12 — full-screen overlay shown when `AppState.requireBiometricAtLaunch`
/// is on and the runtime `isUnlocked` latch is false. Renders a centered
/// brand mark + Face ID button; tapping fires `BiometricGate.evaluate`
/// and on success flips the latch via `AppState.markUnlocked`.
///
/// The view itself doesn't read or write the user preference — that's
/// owned by Settings. This is purely the lock surface.
///
/// Failure modes:
///   - User cancels → stays on lock screen with a "Try again" button.
///   - Biometric not available on device → falls back to a "Sign out
///     of Flagship to continue" affordance (no useful auth path forward
///     from a phone with broken biometrics + Face ID-required setting).
struct BiometricLockScreen: View {
    @Environment(\.colorScheme) private var scheme
    @Environment(AppState.self) private var app
    @State private var status: Status = .idle
    @State private var attemptedAuto = false
    @State private var showStartOver = false

    enum Status: Equatable {
        case idle
        case authenticating
        case failed(String)
    }

    var body: some View {
        let c = FSColors.scheme(scheme)
        ZStack {
            c.bg.ignoresSafeArea()
            VStack(spacing: FS.space.s6) {
                Spacer()
                Image(systemName: "lock.fill")
                    .font(.system(size: 64, weight: .medium))
                    .foregroundColor(c.primary)
                Text("Flagship is locked")
                    .font(.system(size: 22, weight: .semibold))
                    .foregroundColor(c.text)
                Text("Unlock with Face ID to continue.")
                    .font(FS.font.bodySm())
                    .foregroundColor(c.textMuted)
                Spacer()
                if case let .failed(msg) = status {
                    Text(msg)
                        .font(FS.font.caption())
                        .foregroundColor(c.danger)
                        .multilineTextAlignment(.center)
                        .padding(.horizontal, FS.space.s6)
                }
                FSPrimaryButton(
                    status == .authenticating ? "Authenticating…" : "Unlock with Face ID",
                    enabled: status != .authenticating,
                    block: true,
                    large: true,
                ) {
                    Task { await tryUnlock() }
                }
                .padding(.horizontal, FS.space.s6)
                // Always-present escape so a stale Keychain identity (which
                // survives an app delete+reinstall) or broken/absent biometrics
                // can never trap the user on this screen with no way forward.
                // This is a genuine Tier-2 SIGN OUT — it erases this device's
                // local key material from the Keychain (every profile, since
                // the locked user can't pick one) and drops to Welcome. Like
                // the Settings sign-out it's snoop-hardening, not eviction:
                // the same key comes back via passkey recovery (instant
                // re-pair, no rotation). Confirm copy adapts on cloud recovery.
                Button("Sign out") { showStartOver = true }
                    .font(FS.font.bodySm())
                    .foregroundColor(c.textMuted)
                    .padding(.bottom, FS.space.s10)
                    .accessibilityIdentifier("lock-sign-out-btn")
            }
        }
        .accessibilityIdentifier("biometric-lock-screen")
        .confirmationDialog(
            app.hasCloudRecovery ? "Sign out of this device?" : "Sign out without recovery?",
            isPresented: $showStartOver,
            titleVisibility: .visible
        ) {
            Button(
                app.hasCloudRecovery ? "Sign out" : "Sign out anyway",
                role: .destructive
            ) {
                // Tier-2 sign out from the lock screen: erase ALL Keychain
                // profiles (the locked user can't choose one) — the wrapped
                // UMK / IRK that survive an app reinstall — then drop to
                // Welcome. No server-side revoke; the same key returns via
                // passkey recovery.
                Keystore.wipeAllProfiles()
                app.signOut()
            }
            Button("Cancel", role: .cancel) {}
        } message: {
            Text(app.hasCloudRecovery
                ? "This erases this device's account key from the Keychain and returns you to the start. Sign back in with your recovery passkey to restore it — your account and servers stay put."
                : "⚠️ You have NO cloud recovery enrolled. Erasing this device's key with no backup means there's no way to sign back in — your account access is lost for good.")
        }
        // Auto-prompt on first appearance — pattern most password-manager
        // apps follow — EXCEPT when the user reached this screen by tapping
        // "Lock" in Settings. A deliberate lock that instantly auto-unlocks
        // itself is pointless, so in that case we wait for an explicit tap.
        // Launch + return-from-background still auto-prompt. Subsequent failures
        // also require an explicit tap so we don't infinite-loop on "User
        // cancelled" if the user genuinely wants to back out and force-quit.
        .task {
            if !attemptedAuto {
                attemptedAuto = true
                if !app.awaitingManualUnlock {
                    await tryUnlock()
                }
            }
        }
    }

    private func tryUnlock() async {
        status = .authenticating
        do {
            try await BiometricGate().evaluate(reason: "Unlock Flagship")
            app.markUnlocked()
            status = .idle
        } catch BiometricGate.GateError.notAvailable {
            status = .failed("Face ID isn't set up on this device. Disable the lock from Settings on another paired device, or sign out and start fresh.")
        } catch BiometricGate.GateError.userCancelled {
            status = .failed("Cancelled. Tap above to try again.")
        } catch {
            status = .failed("Couldn't authenticate: \(error.localizedDescription)")
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/ui/shell/BiometricLockScreen.kt
// C12 — Compose mirror of FlagshipUI/Shell/BiometricLockScreen.swift.
//
// Renders a centered "Flagship is locked" lockout whenever
// AppState.requireBiometricAtLaunch is on and isUnlocked is false.
// Tapping the Unlock button fires BiometricGate.evaluate against the
// hosting FragmentActivity and on success flips the latch via
// AppState.markUnlocked.

package com.flagshipserver.app.ui.shell

import android.app.Activity
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.outlined.Lock
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.fragment.app.FragmentActivity
import com.flagshipserver.app.core.LocalAppState
import com.flagshipserver.app.core.NetworkErrorHumanizer
import com.flagshipserver.app.keystore.BiometricCancelled
import com.flagshipserver.app.keystore.BiometricGate
import com.flagshipserver.app.keystore.Keystore
import com.flagshipserver.app.ui.components.FSGhostButton
import com.flagshipserver.app.ui.components.FSPrimaryButton
import com.flagshipserver.app.ui.theme.FS
import kotlinx.coroutines.launch

@Composable
fun BiometricLockScreen() {
    val app = LocalAppState.current
    val ctx = LocalContext.current
    val scope = rememberCoroutineScope()
    var status by remember { mutableStateOf<Status>(Status.Idle) }
    var attemptedAuto by remember { mutableStateOf(false) }
    // Tier 2 — drives the lock-screen Sign-out confirmation. Copy +
    // severity adapt on cloud recovery (same framing as Settings).
    var showSignOut by remember { mutableStateOf(false) }
    val hasRecovery by app.hasCloudRecovery.collectAsState()

    val activity = ctx as? FragmentActivity
    Box(
        modifier = Modifier.fillMaxSize().padding(horizontal = FS.space.s6),
        contentAlignment = Alignment.Center,
    ) {
        Column(
            horizontalAlignment = Alignment.CenterHorizontally,
            verticalArrangement = Arrangement.spacedBy(FS.space.s4),
            modifier = Modifier.fillMaxSize(),
        ) {
            Spacer(Modifier.weight(1f))
            Icon(
                Icons.Outlined.Lock,
                contentDescription = null,
                tint = FS.colors.primary,
                modifier = Modifier.size(64.dp),
            )
            Text(
                "Flagship is locked",
                color = FS.colors.text,
                style = TextStyle(fontSize = 22.sp, fontWeight = FontWeight.SemiBold),
            )
            Text(
                "Unlock with biometrics to continue.",
                color = FS.colors.textMuted,
                style = TextStyle(fontSize = 14.sp),
            )
            Spacer(Modifier.weight(1f))
            (status as? Status.Failed)?.let {
                Text(
                    it.message,
                    color = FS.colors.danger,
                    style = TextStyle(fontSize = 12.sp),
                )
                Spacer(Modifier.height(FS.space.s2))
            }
            FSPrimaryButton(
                label = if (status == Status.Authenticating) "Authenticating…" else "Unlock with biometrics",
                onClick = {
                    if (activity == null) {
                        status = Status.Failed("Lost activity context — restart the app to retry.")
                        return@FSPrimaryButton
                    }
                    scope.launch { tryUnlock(activity) { newStatus ->
                        status = newStatus
                        if (newStatus == Status.Idle) app.markUnlocked()
                    } }
                },
                enabled = status != Status.Authenticating,
                block = true,
                large = true,
            )
            Spacer(Modifier.height(FS.space.s4))
            // Always-present escape so broken/absent biometrics can never
            // trap the user on this screen with no way forward. This is a
            // genuine Tier-2 SIGN OUT — it erases this device's local key
            // material from the Keystore (every profile, since the locked
            // user can't pick one) and drops to Welcome. Like the Settings
            // sign-out it's snoop-hardening, not eviction: the same key
            // comes back via passkey recovery (instant re-pair, no
            // rotation). Confirm copy adapts on cloud recovery.
            FSGhostButton(
                label = "Sign out",
                onClick = { showSignOut = true },
                block = true,
            )
            Spacer(Modifier.height(FS.space.s8))
        }
    }
    if (showSignOut) {
        AlertDialog(
            onDismissRequest = { showSignOut = false },
            confirmButton = {
                TextButton(onClick = {
                    showSignOut = false
                    // Tier-2 sign out from the lock screen: erase ALL
                    // Keystore profiles (the locked user can't choose one)
                    // — the UMK / IRK that survive an app reinstall — then
                    // drop to Welcome. No server-side revoke; the same key
                    // returns via passkey recovery.
                    Keystore.wipeAllProfiles()
                    app.signOut()
                }) {
                    Text(
                        if (hasRecovery) "Sign out" else "Sign out anyway",
                        color = FS.colors.danger,
                    )
                }
            },
            dismissButton = {
                TextButton(onClick = { showSignOut = false }) { Text("Cancel") }
            },
            title = {
                Text(
                    if (hasRecovery) "Sign out of this device?"
                    else "Sign out without recovery?",
                )
            },
            text = {
                Text(
                    if (hasRecovery)
                        "This erases this device's account key from the Keystore and returns " +
                            "you to the start. Sign back in with your recovery passkey to " +
                            "restore it — your account and servers stay put."
                    else
                        "⚠️ You have NO cloud recovery enrolled. Erasing this device's key with " +
                            "no backup means there's no way to sign back in — your account " +
                            "access is lost for good.",
                )
            },
        )
    }
    // Auto-prompt once on first appearance — EXCEPT when the user reached this
    // screen by tapping "Lock" in Settings. A deliberate lock that instantly
    // auto-unlocks itself is pointless, so in that case we wait for an explicit
    // tap. Launch + return-from-background still auto-prompt.
    LaunchedEffect(Unit) {
        if (!attemptedAuto && activity != null && !app.awaitingManualUnlock.value) {
            attemptedAuto = true
            tryUnlock(activity) { newStatus ->
                status = newStatus
                if (newStatus == Status.Idle) app.markUnlocked()
            }
        }
    }
}

private sealed class Status {
    data object Idle : Status()
    data object Authenticating : Status()
    data class Failed(val message: String) : Status()
}

private suspend fun tryUnlock(
    activity: FragmentActivity,
    setStatus: (Status) -> Unit,
) {
    setStatus(Status.Authenticating)
    try {
        BiometricGate.evaluate(activity, title = "Unlock Flagship", subtitle = "Use biometrics to continue")
        setStatus(Status.Idle)
    } catch (_: BiometricCancelled) {
        setStatus(Status.Failed("Cancelled. Tap above to try again."))
    } catch (e: Throwable) {
        // UX-B — never leak the raw exception text to the user.
        setStatus(Status.Failed(NetworkErrorHumanizer.humanize(e)))
    }
}
